src/main/scala/state/Terms.scala

```diff
@@ -401,20 +461,120 @@ trait StructuralEqualityBinaryOp[E] extends BinaryOp[E] {
+        p.productIterator.mkString(", ")
+      case trm => trm.toString
+    }
+    s"${this.getClass.getSimpleName}(${argString})"
+  }
+}
 
-  override def equals(other: Any) = (
-    this.eq(other.asInstanceOf[AnyRef])
-      || (other match {
-      case se: StructuralEquality if this.getClass.eq(se.getClass) =>
-        equalityDefiningMembers == se.equalityDefiningMembers
-      case _ => false
-    }))
+trait ConditionalFlyweightBinaryOp[T] extends ConditionalFlyweight[(Term, Term), T] with BinaryOp[Term] with Term {
+  override val equalityDefiningMembers = (p0,  p1)
 }
 
+trait ConditionalFlyweightUnaryOp[T] extends ConditionalFlyweight[Term, T] with UnaryOp[Term] with Term {
+  override val equalityDefiningMembers = p
+}
+
+/**
+  * Version of CondFlyweightFactory where the return type of apply is Term, i.e., the apply method may return
+  * arbitrary terms due to simplification.
+  * @tparam T constructor argument type of the class V (i.e., a tuple or the type of the only argument)
+  * @tparam V class we are creating instances of
+  */
+trait CondFlyweightTermFactory[T, V <: ConditionalFlyweight[T, V] with Term] extends CondFlyweightFactory[T, Term, V]
+
+/**
+  * Version of CondFlyweightFactory where the return type of apply is V, i.e., the apply method always returns an
+  * instance of the class whose apply method is called.
+  * @tparam T constructor argument type of the class V (i.e., a tuple or the type of the only argument)
+  * @tparam V class we are creating instances of
+  */
+trait PreciseCondFlyweightFactory[T, V <: ConditionalFlyweight[T, V] with Term] extends CondFlyweightFactory[T, V, V]
+
+
+/**
+  * Default version of GeneralCondFlyweightFactory where the arguments of the apply method is the same as the
+  * constructor arguments of class V.
+  * @tparam T constructor argument type of the class V (i.e., a tuple or the type of the only argument)
+  * @tparam U return type of the apply method (i.e., either the class itself or something more general for simplifying
+  *           apply methods)
+  * @tparam V class we are creating instances of
+  */
+trait CondFlyweightFactory[T, U, V <: U with ConditionalFlyweight[T, V]] extends GeneralCondFlyweightFactory[T, T, U, V] {
+  def apply(v1: T): U = createIfNonExistent(v1)
+
+}
+
+/**
+  * Most general trait to be implemented by companion objects to create instances of ConditionalFlyweight[T, V].
+  * @tparam IF parameter type of the apply method (can be different from the constructor args of the class)
+  * @tparam T constructor argument type of the class V (i.e., a tuple or the type of the only argument)
+  * @tparam U return type of the apply method (i.e., either the class itself or something more general for simplifying
+  *           apply methods)
+  * @tparam V class we are creating instances of
+  */
+trait GeneralCondFlyweightFactory[IF, T <: IF, U, V <: U with ConditionalFlyweight[T, V]] extends (IF => U) {
+
+  import scala.collection.concurrent.TrieMap
```

Comment:
Personally not a fan of non-top-level imports, except maybe for implicits. Having all imports at the top seems clearer to me.